server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -189,13 +189,16 @@ public void writeTo(StreamOutput out) throws IOException {
         }
     }
 
+    /**
+     * This is a safety limit that should only be exceeded in very rare and special cases. The assumption is that
+     * 99% of the users have less than 1024 shards per index. We also make it a hard check that requires restart of nodes
+     * if a cluster should allow to create more than 1024 shards per index. NOTE: this does not limit the number of shards
+     * per cluster. this also prevents creating stuff like a new index with millions of shards by accident which essentially
+     * kills the entire cluster with OOM on the spot.
+     */
+    public static final String MAX_NUMBER_OF_SHARDS_UPPER_BOUND = "1024";
```

Comment:
Naming discussion: Can we make it explicit that this is about a single index since we have the same max number of shards concept per node? Also `upper_bound` seems redundant? So something like `PER_INDEX_MAX_NUMBER_OF_SHARDS`?